warcprox/mitmproxy.py

```diff
@@ -175,7 +176,7 @@ def begin(self, extra_response_headers={}):
 
         for k,v in self.msg.items():
             if k.lower() not in (
-                    'connection', 'proxy-connection', 'keep-alive',
+                    'connection', 'proxy-connection',
```

Comment:
In your commit comment you wrote:
> Now that we reuse remote connections, its better NOT to remove the `keep-alive` request header. We need to send it to the remote host to make it keep the connection open if possible.

Good point! But I think your code isn't quite doing the right thing. First of all this is the wrong spot. These are the headers from the remote server, and warcprox is filtering out hop-by-hop headers that should not be sent to the proxy client. The request header adjustment happens in `_proxy_request()` around line 408. What you should do there I think is merely add the header `Connection: keep-alive` (although that may not even be necessary, since keep-alive should happen by default).

Reply:
In fact, seems to me the fact that the connection reuse works at all shows that the header probably isn't necessary.

Reply:
Oh, I'm sorry for this error! I'll put back `keep-alive` from where I deleted it.

About the fact that keep alive should happen by default:
According to my knowledge, some http servers do keep alive all connections regardless of the client request headers but others do it only when the client sends the Keep-Alive header.
(Personally, I like to turn it always on in my projects using nginx :))

I see that we delete the header `Keep-Alive` header we send to the remote server in `_proxy_request` line 408.
I tried stopping that, run my benchmark again and saw no difference in the number of new and existing connections in warcprox profiling.
Thus, I conclude that keep-alive happens by default indeed and no other change is needed here.